### ISSUE TITLE
Unable to upload files larger >= 2GB

### ISSUE THREAD

**Issue by agrinh** [dask/gcsfs#325]
I'm experiencing an issue when trying to upload files larger than 2GB. 

**What happened**: I get an overflow error when uploading large files

**What you expected to happen**: Upload to finish successfully

**Minimal Complete Verifiable Example**:

```python
import gcsfs

fs = gcsfs.GCSFileSystem()

# Works fine
with fs.open("gs://bucket/path/to/upload/one", "wb") as f:
    f.write(b"0" * (2 ** 30))

# Raises `ClientOSError: [Errno 1] malloc failure (_ssl.c:2423)`
with fs.open("gs://bucket/path/to/upload/two", "wb") as f:
    f.write(b"0" * (2 ** 31 - 1))

# Raises `OverflowError: string longer than 2147483647 bytes`
with fs.open("gs://bucket/path/to/upload/three", "wb") as f:
    f.write(b"0" * (2 ** 31))
```

**Anything else we need to know?**: I've tried this on various machines and see the same thing each time, definitely not mac only.

**Environment**:

- Python version: 3.7.9
- Operating System: macOS Catalina 10.15.7
- Install method (conda, pip, source): pip
- GCSFS version: 0.7.1

**Comment by martindurant** [comment]
Apparently this is a known issue with aiohttp. You could work around it by splitting your writes into multiple pieces of <2**30bytes.

Our code only currently checks for blocks that are too small, and it seems no one has tried to push a block this big before. The code could be changed to split such big blocks internally to save you the hassle.

**Comment by agrinh** [comment]
> Apparently this is a known issue with aiohttp. You could work around it by splitting your writes into multiple pieces of <2**30bytes.
> 
> Our code only currently checks for blocks that are too small, and it seems no one has tried to push a block this big before. The code could be changed to split such big blocks internally to save you the hassle.

Oh, that's a bit of a surprise. Can you share a link to the known aiohttp issue, please?

Also, is there any chance you could provide some pointers for fixing this? I might be interested in helping out.

**Comment by martindurant** [comment]
Actually, it seems to be deeper, in ssl or httplib, see discussions like https://github.com/psf/requests/issues/2717 (for the requests library).

So there are two possibilities, the first of which might not work, but is far simpler
- in `GCSFile._upload_chunk` passing the internal BytesIO file-like `self.buffer` to the HTTP call, instead of extracting its bytes data (see https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession , in `request`, `data` can be a file-like rather than bytes). Note that the API stated that it's still possible that only part of the file is uploaded.
- wrap the current code in `GCSFile._upload_chunk` to check and break up the bytes block to be below some maximum size, presumably 2**30 (or maybe user settable).

**Comment by martindurant** [comment]
@agrinh : do you think you will be able to make a fix based on this?

**Comment by agrinh** [comment]
> @agrinh : do you think you will be able to make a fix based on this?

It's enough to get me started, let me see if I can get some time to try your first suggestion to begin with.

**Comment by agrinh** [comment]
Fixed in #334
